### ISSUE TITLE
NCL issue with netcdf output, possibly related to >2Gb limit

### ISSUE THREAD

**Issue by tsjackson-noaa** [NOAA-GFDL/MDTF-diagnostics#18]
Submitted via email by Will Cooke (2/4/2020).
"The MJO_suite.log file reports
ncendef: ncid 196608: NetCDF: One or more variable sizes violate format constraints".

This appears to be an issue with NCL, see [link](https://www.ncl.ucar.edu/Support/talk_archives/2013/0966.html) and the [FAQ item](http://www.ncl.ucar.edu/FAQ/#file_io_004).

**Comment by tsjackson-noaa** [comment]
**Progress update:**

I don't have access to the log file Will was quoting, so I can't determine which write operation in MJO_suite caused the error. Will was reporting abnormally long runtimes for MJO_suite, and it's odd that this bug wasn't encountered in the 30-year runs on CM4 and ESM4.

Regardless, this is a serious bug, as PODs should be able to handle long analysis periods.

The conda-based installation installs a current version of NCL and the netCDF4 libraries, so we would like to globally change that NCL's default netCDF format from "Classic" to "NetCDF4Classic". Using `ncl -f` [might](https://www.ncl.ucar.edu/Document/Manuals/Ref_Manual/NclCLO.shtml) accomplish this, but I can't find further documentation on the details of what this flag does. The alternative would be to go into every NCL-based POD's code and insert a [setfileoption](http://www.ncl.ucar.edu/Document/Functions/Built-in/setfileoption.shtml) command every time a netCDF file is written.

Reminder on the differences between netCDF formats: [1](https://www.unidata.ucar.edu/software/netcdf/docs/faq.html#How-many-netCDF-formats-are-there-and-what-are-the-differences-among-them), [2](https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_introduction.html#netcdf_format).

**Comment by jkrasting** [comment]
@tsjackson-noaa and @wfcooke - thanks for finding this issue.  SPEAR creates some great use cases that help uncover new issues.  We were making some progress here, but COVID-19 slowed things down.

Does this POD fail or just run slowly on SPEAR data?

**Comment by tsjackson-noaa** [comment]
I'll try to dig through the emails we exchanged with Will to see if I can get any more details about this run. Like I said, I haven't hit this error, and I don't think Will was throwing a large amount of model data at the framework in his tests.

@Wen-hao-Dong confirmed that `ncl -f` will enable files to handle > 2gb by default, with the caveat that files seem to be limited to 4gb if the time axis is not unlimited (http://www.doc.ic.ac.uk/~rn710/Installs/netcdf-4.2/man4/html/offset_format_limitations.html).

My plan to fix this issue is as follows. Currently the NCL PODs call the NCL scripts from a top-level python script, using a copy of the following code from v2.0 of the framework, eg, https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/170226663a52d05fb02f9f6d5c426e6042317a5b/diagnostics/EOF_500hPa/EOF_500hPa.py#L15-L32
I'll create a new `/shared` module containing code for use by both the framework and PODs, which will have a single copy of `generate_ncl_plots` that uses the `-f` flag. The top-level scripts of the NCL pods will then import the function from there.

**Comment by tsjackson-noaa** [comment]
It appears that invoking ncl with `-f` is doing other things I don't understand. I started with a data set that MJO_suite ran successfully on, but when I changed `ncl` to `ncl -f` in the python wrapper script (keeping the data and all other input the same) I hit the following error:

```
(0)     daily_netcdf.ncl reading /vftmp/Thomas.Jackson/pid40305/wkdir/MDTF_CM4_piControl_v5_proto1_1_10/MJO_suite/day/CM4_piControl_v5_proto1.pr.day.nc for making precip file!
(0)     Interpolation: Stored lat/lon from pr as desired resolution160 576
fatal:["NclNetCDF4.c":5409]:NetCDF: Operation not allowed in data mode
fatal:["NclAdvancedFile.c":6128]:AdvancedFileAddVar: Error adding variable <date> as type: <none>
fatal:["Execute.c":8637]:Execute: Error occurred at or near line 312 in file $POD_HOME/utils.ncl

fatal:["Execute.c":8637]:Execute: Error occurred at or near line 135 in file /home/Thomas.Jackson/mdtf/MDTF-diagnostics/diagnostics/MJO_suite/daily_netcdf.ncl
```

**Comment by Wen-hao-Dong** [comment]
> It appears that invoking ncl with `-f` is doing other things I don't understand. I started with a data set that MJO_suite ran successfully on, but when I changed `ncl` to `ncl -f` in the python wrapper script (keeping the data and all other input the same) I hit the following error:
> 
> ```
> (0)     daily_netcdf.ncl reading /vftmp/Thomas.Jackson/pid40305/wkdir/MDTF_CM4_piControl_v5_proto1_1_10/MJO_suite/day/CM4_piControl_v5_proto1.pr.day.nc for making precip file!
> (0)     Interpolation: Stored lat/lon from pr as desired resolution160 576
> fatal:["NclNetCDF4.c":5409]:NetCDF: Operation not allowed in data mode
> fatal:["NclAdvancedFile.c":6128]:AdvancedFileAddVar: Error adding variable <date> as type: <none>
> fatal:["Execute.c":8637]:Execute: Error occurred at or near line 312 in file $POD_HOME/utils.ncl
> 
> fatal:["Execute.c":8637]:Execute: Error occurred at or near line 135 in file /home/Thomas.Jackson/mdtf/MDTF-diagnostics/diagnostics/MJO_suite/daily_netcdf.ncl
> ```

The main difference in the saved netcdf file from _ncl -f_ and _ncl_ command is the attributes for the saved variables will be omitted if the _ncl -f_ command is used. At line 312 in file $POD_HOME/utils.ncl:

   gw!0 = "lat"
   gw&lat = var&lat
   fo->gw = gw
   date!0 = "time"
   **date&time = var&time** (Line 312)
   fo->date = date

There is no attribute in var&time when _ncl -f_ is used

BTW, we can get rid of the _ncl -f_ command by adding one sentence in the script    

 **setfileoption("nc","Format","NetCDF4Classic")**

This allows us to save a variable larger than 2GB gracefully. 

Or **setfileoption("nc","Format","NetCDF4")** for variable larger than 4GB

Please see more [here](https://www.ncl.ucar.edu/Document/Functions/Built-in/setfileoption.shtml )

**Comment by tsjackson-noaa** [comment]
Thanks so much for figuring out my problem so quickly, @Wen-hao-Dong ! 

I thought the `-f` flag did the same thing as the setfileoption() command -- I wanted to use the flag because it's messier to have to insert a setfileoption() command in every NCL script (plus we'd have to remember to do this on any new NCL scripts), but it looks like this is the way to resolve this issue.

We currently save files provided as input to the scripts as netcdf4 classic (I forget what "backwards compatibility" issues I ran into in 41b68866367aebe12339100ab273422148ae65c2 ; maybe it was this stuff) but I think I'll put the format used by both the framework and setfileoption() in an environment variable in case we need to switch to netcdf4 for really big data in the future.

**Comment by tsjackson-noaa** [comment]
Closing this issue under the assumption that our needs have been addressed by PR #263 (for now at least).
